class Buffer contains defined behaviour (using delete[] against a void pointer)

Running Vanilla Conquer with TD has highlighted memory warnings upon start-up

![image](https://user-images.githubusercontent.com/10485456/109120310-b0ed8a80-773d-11eb-851c-a8096f5a3d24.png)

This is because BufferPtr is declared as a void*, and  doing a delete[] on a void* is undefined in C++

https://discord.com/channels/409121752921276426/770995500677333002/814040181668577280

A simple helper program demonstrates this

int main ()
{
    void *buf;
    buf = new char[4096];
    delete[] buf;
    return 0;
}
main.cc:7:11: warning: deleting ‘void*’ is undefined [-Wdelete-incomplete]
    7 |  delete[] buf;
      |           ^~~
and also reproduces the valgrind warning
==114685== Mismatched free() / delete / delete []
==114685==    at 0x483CFBF: operator delete(void*) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==114685==    by 0x109193: main (in /home/terrorcide/main)
==114685==  Address 0x4dbcc80 is 0 bytes inside a block of size 4,096 alloc'd
==114685==    at 0x483C583: operator new[](unsigned long) (in /usr/lib/x86_64-linux-gnu/valgrind/vgpreload_memcheck-amd64-linux.so)
==114685==    by 0x10917E: main (in /home/terrorcide/main)

--- a/common/buff.cpp
+++ b/common/buff.cpp
@@ -152,7 +152,7 @@ Buffer& Buffer::operator=(Buffer const& buffer)
 {
     if (buffer != this) {
         if (IsAllocated) {
-            delete[] BufferPtr;
+            delete[] static_cast<char*>(BufferPtr);
         }
         IsAllocated = false;
         BufferPtr = buffer.BufferPtr;
@@ -200,7 +200,7 @@ Buffer::~Buffer(void)
 void Buffer::Reset(void)
 {
     if (IsAllocated) {
-        delete[] BufferPtr;
+        delete[] static_cast<char*>(BufferPtr);
     }
     BufferPtr = NULL;
     Size = 0;

--- a/common/linear.cpp
+++ b/common/linear.cpp
@@ -190,7 +190,7 @@ bool Linear_Scale_To_Linear(void* thisptr,
                 int xrat = 0;
 
                 for (int j = 0; j < dst_x1; ++j) {
-                    char tmp = s[xrat >> 16];
+                    unsigned char tmp = s[xrat >> 16];
 
                     if (tmp != 0) {
                         *d = (remap)[tmp];
@@ -207,7 +207,7 @@ bool Linear_Scale_To_Linear(void* thisptr,
                 int xrat = 0;
 
                 for (int j = 0; j < dst_x1; ++j) {
-                    char tmp = s[xrat >> 16];
+                    unsigned char tmp = s[xrat >> 16];
 
                     if (tmp != 0) {
                         *d = tmp;

--- a/common/load.cpp
+++ b/common/load.cpp
@@ -166,7 +166,7 @@ void* Load_Alloc_Data(const FileClass& file)
     void* ptr = 0;
     long size = const_cast<FileClass&>(file).Size();
 
-    ptr = new char[size];
+    ptr = malloc(size);
     if (ptr) {
         const_cast<FileClass&>(file).Read(ptr, size);
     }

--- a/common/mixfile.h
+++ b/common/mixfile.h
@@ -207,7 +207,7 @@ template <class T> MixFileClass<T>::~MixFileClass(void)
         free((char*)Filename);
     }
     if (Data != NULL && IsAllocated) {
-        delete[] Data;
+        delete[] static_cast<char*>(Data);
         IsAllocated = false;
     }
     Data = NULL;

--- a/tiberiandawn/bdata.cpp
+++ b/tiberiandawn/bdata.cpp
@@ -4301,7 +4301,7 @@ short const* BuildingTypeClass::Occupy_List(bool placement) const
         }
 
         SmudgeTypeClass const& smudge = SmudgeTypeClass::As_Reference(bib);
-        static short _list[25];
+        static short _list[50];
         short* dest = &_list[0];
 
         /*

--- a/tiberiandawn/display.cpp
+++ b/tiberiandawn/display.cpp
@@ -734,8 +734,14 @@ void DisplayClass::Set_Cursor_Shape(short const* list)
     if (list) {
         int w, h;
         static short _list[50];
+        const short* src = list;
+        short* dest = _list;
 
-        memcpy(_list, list, sizeof(_list));
+        while (*src != REFRESH_EOL) {
+            *dest++ = *src++;
+        }
+        *dest = REFRESH_EOL;
+
         CursorSize = _list;
         Get_Occupy_Dimensions(w, h, CursorSize);
         ZoneOffset = -(((h / 2) * MAP_CELL_W) + (w / 2));

--- a/tiberiandawn/display.cpp
+++ b/tiberiandawn/display.cpp
@@ -541,7 +541,7 @@ void DisplayClass::Init_Theater(TheaterType theater)
  *=============================================================================================*/
 short const* DisplayClass::Text_Overlap_List(char const* text, int x, int y, int lines)
 {
-    static short _list[30];
+    static short _list[50];
 
     if (text) {
         short* ptr = &_list[0];

--- a/tiberiandawn/drive.cpp
+++ b/tiberiandawn/drive.cpp
@@ -1244,12 +1244,12 @@ bool DriveClass::Start_Of_Move(void)
                             return (true);
                         }
                     } else {
-                        memcpy(&Path[0], &Path[2], CONQUER_PATH_MAX - 2);
+                        memmove(&Path[0], &Path[2], CONQUER_PATH_MAX - 2);
                         Path[CONQUER_PATH_MAX - 2] = FACING_NONE;
                         IsPlanningToLook = true;
                     }
                 } else {
-                    memcpy(&Path[0], &Path[1], CONQUER_PATH_MAX - 1);
+                    memmove(&Path[0], &Path[1], CONQUER_PATH_MAX - 1);
                 }
                 Path[CONQUER_PATH_MAX - 1] = FACING_NONE;
             }

--- a/tiberiandawn/help.cpp
+++ b/tiberiandawn/help.cpp
@@ -50,7 +50,7 @@
 **	data segment. It will be filled in by the Set_Text() function.
 */
 // short const HelpClass::OverlapList[30] = {  // Can't be const - it's expected to be written to. ST - 2/7/2019 5:16PM
-short HelpClass::OverlapList[30] = {REFRESH_EOL};
+short HelpClass::OverlapList[50] = {REFRESH_EOL};
 
 char const* HelpClass::HelpText;
 

--- a/tiberiandawn/help.h
+++ b/tiberiandawn/help.h
@@ -135,7 +135,7 @@ private:
     **	the help text is to be erased.
     */
     // static short const OverlapList[30];
-    static short OverlapList[30]; // Can't be const - it's expected to be written to. ST - 2/7/2019 5:16PM
+    static short OverlapList[50]; // Can't be const - it's expected to be written to. ST - 2/7/2019 5:16PM
 
     enum HelpClassEnum
     {

--- a/tiberiandawn/mapsel.cpp
+++ b/tiberiandawn/mapsel.cpp
@@ -1130,8 +1130,8 @@ void Map_Selection(void)
     Theme.Queue_Song(THEME_NONE);
     Fade_Palette_To(BlackPalette, FADE_PALETTE_MEDIUM, NULL);
     delete europe;
-    delete progresspalette;
-    delete grey2palette;
+    delete[] progresspalette;
+    delete[] grey2palette;
     delete TextPrintBuffer;
     TextPrintBuffer = NULL;
     BlitList.Clear();

--- a/tiberiandawn/sidebar.cpp
+++ b/tiberiandawn/sidebar.cpp
@@ -2096,7 +2096,7 @@ bool SidebarClass::StripClass::Recalc(void)
             **	Removes this entry from the list.
             */
             if (BuildableCount > 1 && index < BuildableCount - 1) {
-                memcpy(
+                memmove(
                     &Buildables[index], &Buildables[index + 1], sizeof(Buildables[0]) * ((BuildableCount - index) - 1));
             }
             TopIndex = 0;

--- a/tiberiandawn/sounddlg.cpp
+++ b/tiberiandawn/sounddlg.cpp
@@ -449,7 +449,7 @@ void SoundControlsClass::Process(void)
     while (listbox.Count()) {
         char const* ptr = listbox.Get_Item(0);
         listbox.Remove_Item(ptr);
-        delete[](void*) ptr;
+        delete[](char*) ptr;
     }
 }
 

--- a/tiberiandawn/techno.cpp
+++ b/tiberiandawn/techno.cpp
@@ -3918,7 +3918,7 @@ void TechnoClass::Base_Is_Attacked(TechnoClass const* enemy)
                         defender[lp] = (FootClass*)infantry;
                         continue;
                     }
-                    if (value[count] < newweakest) {
+                    if (count < 6 && value[count] < newweakest) {
                         newweakest = value[lp];
                     }
                 }
@@ -3980,7 +3980,7 @@ void TechnoClass::Base_Is_Attacked(TechnoClass const* enemy)
                         defender[lp] = (FootClass*)unit;
                         continue;
                     }
-                    if (value[count] < newweakest) {
+                    if (count < 6 && value[count] < newweakest) {
                         newweakest = value[lp];
                     }
                 }

